### ISSUE TITLE
Python runtime exception while running learn_static::start_learning(): ValueError: cannot convert float NaN to integer

### ISSUE THREAD

**Issue by jordi-torrents** [mooch443/trex#230]
Hi Tristan, I’m encountering a NaN validation loss in TRex.

I processed two different videos of zebrafish using Tgrabs on two separate Linux machines, testing both TRex versions 1.1.9 and 1.1.8. In all cases, I achieved good segmentation results, with the animals clearly visible and coherent base identities. However, when running the "Visual Identification" feature in TRex, the process fails after the validation of the first epoch, displaying a NaN validation loss error.

I'm using the default parameters, modifying only the Tgrabs threshold, the `track_max_individuals` and a `track_include`.

 - Operating system: Ubuntu 24.04
 - Build hash: 4ce4be1cd1679bcca4da5af8780dd86c1d31050b
 - Terminal used: Bash

I'm attaching the log file, [trex.log](https://github.com/user-attachments/files/17996033/trex.log). Thanks for your time!

**Comment by mooch443** [comment]
Hey, thanks for your report! 

It's difficult to say from here - but one thing is that it's quite a low number of images per individual. So <200, I think all of them. Not sure that's the problem, and it shouldn't produce NaN anyway, but certainly doesn't help.

If the base tracking looks fine to you, maybe you can check the tracklet lengths: select an individual and check the list of segments/trajectory pieces on the top left. If they are all quite short then we'd have to figure out what causes it. Maybe it's the best one can do, but in many cases it helps to increase track_max_speed a bit. This is so it doesn't cause mistrust due to "high speeds". If that doesn't help, maybe your individuals are a bit densely packed?

I can also check out the video if you like! I could run it through the current beta version and see if my rewrite of the visual identification doesn't produce NaN anymore :) 

-Tristan

**Comment by jordi-torrents** [comment]
I tried again with your recommendations. With a video of 10 zebrafish (extracted from [here](https://drive.google.com/drive/folders/1oGpHsu2EfWqU-mSMhO3iAtEKCpcoPIT4)) I got a first training dataset of ~350 images per class, these are the log lines for the first epoch which returns a NaN validation loss:

```
Python received (3473, 80, 80, 1) training images ((3473, 10)) and  (1164, 80, 80, 1) validation images ((1164, 10))
[14:03:30] [py] {'epochs': 150, 'batch_size': 64, 'move_range': 0.025, 'output_path': '/home/jordi/Videos/zebrafish_10_1_weights',
'image_width': 80.0, 'image_height': 80.0, 'classes': array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]), 'learning_rate': 9.999999747378752e-05,
'accumulation_step': 0, 'global_segment': array([14365, 14818]), 'per_epoch': 64, 'min_iterations': 100, 'verbosity': 1}
[14:03:30] [py] # [init] samples per class {0: 346, 1: 340, 2: 341, 3: 342, 4: 359, 5: 348, 6: 340, 7: 345, 8: 359, 9: 353}
[14:03:30] [py] # [init] weights per class {0: 346, 1: 340, 2: 341, 3: 342, 4: 359, 5: 348, 6: 340, 7: 345, 8: 359, 9: 353}
[14:03:30] [py] # [training] data shapes: train (3473, 80, 80, 1) (3473, 10) test (1164, 10) [0 1 2 3 4 5 6 7 8 9]
[14:03:30] [py] # [values] X:142 - 0
[14:03:30] [py] # [values] Y:1.0 - 0.0
[14:03:30] [py] tf.data.Dataset: <ShuffleDataset shapes: ((None, 80, 80, 1), (None, 10)), types: (tf.float32, tf.float32)>
[14:03:30] [py] tf.data.Dataset (validation): <BatchDataset shapes: ((None, 80, 80, 1), (None, 10)), types: (tf.float32, tf.float32)>
Epoch 1/150
64/64 [==============================] - 5s 30ms/step - loss: 0.4755 - accuracy: 0.0999 - val_loss: nan - val_accuracy: 0.0997
```

and then it crashes with the same error as before.

I tried the beta version and the error seems to be fixed in there, but for a publicatin I guess I should not be using that beta version. What standard version do you think I can make work?

**Comment by mooch443** [comment]
> I tried again with your recommendations. With a video of 10 zebrafish (extracted from [here](https://drive.google.com/drive/folders/1oGpHsu2EfWqU-mSMhO3iAtEKCpcoPIT4)) I got a first training dataset of ~350 images per class, these are the log lines for the first epoch which returns a NaN validation loss:
> 
> 
> 
> ```
> 
> Python received (3473, 80, 80, 1) training images ((3473, 10)) and  (1164, 80, 80, 1) validation images ((1164, 10))
> 
> [14:03:30] [py] {'epochs': 150, 'batch_size': 64, 'move_range': 0.025, 'output_path': '/home/jordi/Videos/zebrafish_10_1_weights',
> 
> 'image_width': 80.0, 'image_height': 80.0, 'classes': array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]), 'learning_rate': 9.999999747378752e-05,
> 
> 'accumulation_step': 0, 'global_segment': array([14365, 14818]), 'per_epoch': 64, 'min_iterations': 100, 'verbosity': 1}
> 
> [14:03:30] [py] # [init] samples per class {0: 346, 1: 340, 2: 341, 3: 342, 4: 359, 5: 348, 6: 340, 7: 345, 8: 359, 9: 353}
> 
> [14:03:30] [py] # [init] weights per class {0: 346, 1: 340, 2: 341, 3: 342, 4: 359, 5: 348, 6: 340, 7: 345, 8: 359, 9: 353}
> 
> [14:03:30] [py] # [training] data shapes: train (3473, 80, 80, 1) (3473, 10) test (1164, 10) [0 1 2 3 4 5 6 7 8 9]
> 
> [14:03:30] [py] # [values] X:142 - 0
> 
> [14:03:30] [py] # [values] Y:1.0 - 0.0
> 
> [14:03:30] [py] tf.data.Dataset: <ShuffleDataset shapes: ((None, 80, 80, 1), (None, 10)), types: (tf.float32, tf.float32)>
> 
> [14:03:30] [py] tf.data.Dataset (validation): <BatchDataset shapes: ((None, 80, 80, 1), (None, 10)), types: (tf.float32, tf.float32)>
> 
> Epoch 1/150
> 
> 64/64 [==============================] - 5s 30ms/step - loss: 0.4755 - accuracy: 0.0999 - val_loss: nan - val_accuracy: 0.0997
> 
> ```
> 
> 
> 
> and then it crashes with the same error as before.
> 
> 
> 
> I tried the beta version and the error seems to be fixed in there, but for a publicatin I guess I should not be using that beta version. What standard version do you think I can make work? 

Hey, it's probably fixed there because I'm not using tensorflow anymore :/ the last public version of v119 should work theoretically - can you send me the settings you're using, so I can see whether I can fix it on my side and then let you know?

The samples look fine, so it's a bit confusing that it does that :( 

Thanks,
Tristan

**Comment by jordi-torrents** [comment]
I solved the issue by upgrading tensorflow to 2.10.0, but it's tricky since TRex 1.1.9 runs on Python 3.7 and tensorflows only officially supports PyPI, but not with Python 3.7 and Cuda... Long story short, my receipe to install a working version of TRex is:

Initialize the environment with Python 3.7.*
```
conda create -n trex_env python=3.7
conda activate trex_env
```
Install tensorflow 2.10 from conda-forge since the official pip installation doesn't work with Py3.7 and Cuda
```
conda install tensorflow-gpu -c conda-forge
```
Install trex with all dependencies. Since there is already a tensorflow installed, it's gonna install TRex 1.1.6
```
conda install trex -c trexing
```
Upgrade TRex to latest 1.1.9 without modifying the dependencies
```
conda install trex=1.1.9 -c trexing --no-deps 
```
✨

**Comment by mooch443** [comment]
Very nice! Luckily this won't be necessary anymore soon. :) sorry that you had to go do this for some reason. Tensorflow bug I guess?
